docs/src/abstract_types.md

```diff
@@ -0,0 +1,165 @@
+# Abstract types
+
+AbstractAlgebra.jl implements a couple of abstract types which can be extended.
+
+## Parents
+
+```@setup parents
+using Kroki
+
+parents_bad = plantuml"""
+@startwbs
+skinparam monochrome true
+skinparam arrowThickness 3
+skinparam WbsType tree
+<style>
+wbsDiagram {
+  FontName Monospaced
+  FontSize 16
+  RoundCorner 10
+  Padding 5
+  Margin 15
+}
+</style>
+
++ Set
+++ Ideal{T}
+++ Group
+--- AbstractPermutationGroup
+++ AdditiveGroup
+--- Module{T}
+++++ MatSpace{T}
+---- FPModule{T}
+++ NCRing
++++ NCPolyRing{T}
+--- MatAlgebra{T}
+--- Ring
+++++ PolyRing{T}
+++++ LaurentPolynomialRing{T}
+++++ SeriesRing{T}
+++++ ResRing{T}
+++++ Field
++++++ ResField{T}
++++++ FracField{T}
++++++ NumField{T}
+++++++ SimpleNumField{T}
++++++ FinField
+@endwbs
+"""
+
+parents = plantuml"""
+@startuml
+skinparam monochrome true
+skinparam defaultFontName Monospaced
+skinparam defaultFontSize 16
+skinparam objectArrowColor DarkGray
+skinparam RoundCorner 15
+
+' Ideals
+"Set" -down-> "Ideal{T}"
+
+' Groups
+"Set" -down-> "Group"
+"Group" -down-> "AbstractPermutationGroup"
+
+' Additive groups
+"Set" -down-> "AdditiveGroup"
+"AdditiveGroup" -down-> "Module{T}"
+"Module{T}" -down-> "FPModule{T}"
+"Module{T}" -down-> "MatSpace{T}"
+
+' Non-commutative rings
+"Set" -down-> "NCRing"
+"NCRing" -down--> "MatAlgebra{T}"
+"NCRing" -down--> "NCPolyRing{T}"
+"NCRing" -down-> "Ring"
+
+' Rings
+"Ring" -down-> "PolyRing{T}"
+"PolyRing{T}" -down-> "LaurentPolynomialRing{T}"
+"PolyRing{T}" -down-> "SeriesRing{T}"
+"Ring" -down-> "ResRing{T}"
+"Ring" -down---> "Field"
+
+' Fields
+"Field" -down-> "ResField{T}"
+"Field" -down-> "FracField{T}"
+"Field" -down-> "NumField{T}"
+"NumField{T}" -down-> "SimpleNumField{T}"
+"Field" -down-> "FinField"
+
+hide members
+hide circle
+
+@enduml
+"""
+```
+```@example parents
+parents #hide
+```
+
+## Elements
+
+```@setup elements
+using Kroki
+
+elements = plantuml"""
+@startuml
+skinparam monochrome true
+skinparam defaultFontName Monospaced
+skinparam defaultFontSize 16
+skinparam objectArrowColor DarkGray
+skinparam RoundCorner 15
+
+
+' Ideals
+"SetElem" -down-> "IdealElem{T}"
+
+' Maps
+' NOTE: Identity is 
+"SetElem" -down----> "Map{D, C, S, T}"
+"Map{D, C, S, T}" .right. "SetMap"
+"SetMap" -down-> "IdentityMap"
+"SetMap" -down-> "FunctionalMap"
+"FunctionalMap" -down-> "FPModuleHomomorphism"
+
+' Groups
+' NOTE: Shouldn't AbstractPerm have the suffix -Elem?
```

Comment:
Hmmm... 

But it's not an "an element of a perm(utation)". It could be called "an element of a permutation group", I guess, so "AbstractPermGroupElem" or so, but isn't that just a verbose way to say "a permutation"? From that POV the current name makes a lot of sense...

Then again of course AA already has PolyElem and MatElem -- names I in turn find illogical.... but PermElem would indeed fit with them.

My guess is that the general idea is to just always add the suffix "Elem" ? I see one potential benefit in such a regular strategy: making it "obvious" that a given type indicates elements (although I'd wonder whether this is really necessary for Perm, Poly, Mat?). But then why stop there, and not also extend this regularity to exhibit the correspondence between ring types and element types? For example, it is Set and SetElem; Ring and RingElem;  so why not PolyRing and PolyRingElem (not PolyElem), perhaps with an alias `Poly` (or maybe Poly is the type and PolyRingElem the alias). Similar for MatElem -> MatSpaceElem. 

Of course these names are a bit clunky. But so is PermElem....

I guess there is no perfect answer, just different compromises:-)

Reply:
Thank you for your answer! Yes, I realized that the naming convention isn't always consistent. But that's probably for the best, as you say.

To clarify, I am totally fine with the name `AbstractPerm`. I just thought I'd bring it up in case it was a typo! :slightly_smiling_face:

Reply:
@fingolfin AbstractAlgebra also has Poly and Mat. We can't name all the types the same thing, hence PolyElem and MatElem.